src/Symfony/Component/Finder/Iterator/FilterIterator.php

```diff
@@ -22,12 +22,12 @@
 {
     public function hasChildren()
     {
-        return $this->getInnerIterator() instanceof \RecursiveIterator && $this->getInnerIterator()->hasChildren();
+        return $this instanceof RecursiveAwareIterator && $this->getInnerIterator() instanceof \RecursiveIterator && $this->getInnerIterator()->hasChildren();
```

Comment:
this is a good reason to avoid implementing ``RecursiveIterator`` in our FilterIterator class and putting the implementation of this interface in each filter instead (i.e. only ExcludeDirectoryFilterIterator for now)